mentat/session.py

```diff
@@ -116,24 +119,37 @@ def start(self) -> asyncio.Task[None]:
         async def run_main():
             try:
                 await self._main()
-                await self.stop()
-            except asyncio.CancelledError:
+            except (SessionExit, CancelledError):
                 pass
+            except (MentatError, UserError) as e:
+                self.stream.send(str(e), color="red")
             except Exception:
-                traceback.print_exc()
+                # All unhandled exceptions end up here
+                self.stream.send(
+                    f"Unhandled Exception: {traceback.format_exc()}", color="red"
+                )
+            finally:
+                await self.stop()
 
         setup_logging()
         self._main_task: Task[None] = asyncio.create_task(run_main())
         return self._main_task
 
     async def stop(self):
+        if self.stopped:
+            return
+        self.stopped = True
+
         session_context = SESSION_CONTEXT.get()
         cost_tracker = session_context.cost_tracker
 
         cost_tracker.display_total_cost()
         logging.shutdown()
         self._main_task.cancel()
-        await self._main_task
+        try:
+            await self._main_task
+        except CancelledError:
```

Comment:
When does `await self._main_task` raise a `CancelledError`?

Reply:
We cancel it on the line before this, so I'm pretty sure that's why it raises a CancelledError.

Reply:
Ah that would make sense haha